torch/prototypes/nestedtensor/nested.py

```diff
@@ -0,0 +1,127 @@
+import torch
+
+
+def is_nested_tensor(obj):
+    return isinstance(obj, NestedTensor)
+
+
+def make_nested_tensor(obj):
+    if is_nested_tensor(obj):
+        return obj.clone().detach()
+    elif torch.is_tensor(obj):
+        return obj.clone().detach()
+    else:
+        assert isinstance(obj, list)
+        if len(obj) == 0:
+            return NestedTensor([])
+        for obj_ in obj:
+            assert(torch.is_tensor(obj_))
+        tensors = []
+        for obj_ in obj:
+            tensors.append(obj_.clone().detach())
+        return NestedTensor(tensors)
+
+
+class NestedTensor():
+    def __init__(self, tensors):
+        for tensor in tensors:
+            assert torch.is_tensor(tensor)
+        self.tensors = tensors
+        if len(tensors):
+            self.dim = tensors[0].dim()
+            self.layout = tensors[0].layout
+            self.device = tensors[0].device
+            self.dtype = tensors[0].dtype
+            for tensor in tensors:
+                if not (self.dim == tensor.dim() and
+                        self.layout == tensor.layout and
+                        self.device == tensor.device and
+                        self.dtype == tensor.dtype):
+                    raise ValueError("Each passed Tensor "
+                            "must match in dim, layout, "
+                            "device and dtype")
+        else:
+            empty_tensor = torch.Tensor([])
+            self.dim = empty_tensor.dim()
+            self.layout = empty_tensor.layout
+            self.device = empty_tensor.device
+            self.dtype = empty_tensor.dtype
+
+    def __getattribute__(self, attr):
+        if attr == 'shape':
+            raise NotImplementedError()
+        return super().__getattribute__(attr)
+
+    def __len__(self):
+        return len(self.tensors)
+
+    def __bool__(self):
+        # Explicitly punt on this until we have fully
+        # specificed reduction semantics.
+        raise NotImplementedError()
+
+    def __str__(self):
```

Comment:
This looks the same as `__repr__`. Also, I think on regular tensors these are both the same as well.

Reply:
You're right. That's an unfortunate inconsistency within torch.Tensor. I'll still maintain the distinction in case we fix this. __repr__ should always return a string that can effectively be passed into eval(). This will fail for large tensors...